<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>01 Find empty droplets – Sterile inflammation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-e26003cea8cd680ca0c55a263523d882.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-79108a0fc1995748cbd19a5b0e3e3e7c.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="../site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="../site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Sterile inflammation</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-analysis" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-analysis">    
        <li>
    <a class="dropdown-item" href="../analysis/01_empty_droplets.html">
 <span class="dropdown-text">Empty droplets</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#barcode-ranks" id="toc-barcode-ranks" class="nav-link" data-scroll-target="#barcode-ranks">Barcode ranks</a></li>
  <li><a href="#empty-drops" id="toc-empty-drops" class="nav-link" data-scroll-target="#empty-drops">Empty drops</a></li>
  <li><a href="#comparison" id="toc-comparison" class="nav-link" data-scroll-target="#comparison">Comparison</a></li>
  <li><a href="#selection" id="toc-selection" class="nav-link" data-scroll-target="#selection">Selection</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a>
  <ul class="collapse">
  <li><a href="#parameters" id="toc-parameters" class="nav-link" data-scroll-target="#parameters">Parameters</a></li>
  <li><a href="#output-files" id="toc-output-files" class="nav-link" data-scroll-target="#output-files">Output files</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information">Session information</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">01 Find empty droplets</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>In this document we are going to read in the unfiltered counts matrix produced by <code>Cell Ranger</code> and determine which of those droplets contain cells using the <code>DropletUtils</code> package.</p>
<p>The raw object now contains the raw counts and metadata, with an additional column (CellRangerFilt) marking which cells passed Cell Ranger’s initial filtering. The raw dataset has 32285 features and 2068435 droplets at this point.</p>
<p>The raw object is now pruned to exclude:</p>
<pre><code>Genes that are not expressed in any cell.
Droplets that do not contain any counts (empty droplets).</code></pre>
<p>After removing all zero features and droplets the dataset has 25634 features and 1445128 droplets.</p>
</section>
<section id="barcode-ranks" class="level1">
<h1>Barcode ranks</h1>
<p>Now we will compute barcode rank statistics and identify the knee and inflection points on the total count curve. All detected barcodes are ranked based on their total UMI counts. Barcodes with higher UMI counts (indicative of more RNA transcripts) are assigned higher ranks. Cell-associated barcodes typically have higher UMI counts. Background or empty droplet barcodes have low UMI counts.</p>
<p>Let’s start by ordering the droplets according to their total counts and plotting this on a log scale. This lets us see the distribution of total counts.</p>
<p>The shape of the barcode rank plot provides insights into the quality of the sample. A “cliff-and-knee” shape indicates a good distinction between cell-associated and background barcodes. The steep drop (cliff) represents a clear separation, while the plateau (knee) shows the presence of low-count background barcodes. If the plot lacks this clear structure, it may suggest poor sample quality or high background noise, which can complicate downstream analyses</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="01_empty_droplets_files/figure-html/barcodes-plot-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
<p>This is typical of what we see for 10x experiment where there is a sharp drop off between droplets with lots of counts and those without many. The inflection and knee points are methods for identifying the transition between distributions. These are roughly associated with the cells selected by <code>Cell Ranger</code>. The empty threshold line indicates the point at which we assume droplets must be empty (total counts &lt;= 100).</p>
</section>
<section id="empty-drops" class="level1">
<h1>Empty drops</h1>
<p>We will now look at identifying which droplets to select using the <code>EmptyDrops</code> method. This method tests whether the composition of a droplet is significantly different from the ambient RNA in the sample which is obtained by pooling the empty droplets. Droplets with very large counts are also automatically retained.</p>
<p><code>EmptyDrops</code> uses a permutation approach to calculate p-values, assessing the likelihood that a droplet’s RNA content could arise from ambient RNA alone. Ensuring a sufficient number of iterations (permutations) is vital for accurate p-value estimation. If there are any droplets that have non-significant p-values but are limited by the number of permuations, the number should be increased.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>       Significant
Limited   FALSE    TRUE
  FALSE 1399085    3355
  TRUE      224   42464</code></pre>
</div>
</div>
<p>Limited: Indicates whether the p-value calculation for a droplet was limited by the number of permutations (TRUE = limited, FALSE = not limited). Significant: Indicates whether a droplet was deemed significant (i.e., likely to contain a cell) based on the FDR threshold (TRUE = significant, FALSE = not significant).</p>
<p>Another way to check the <code>EmptyDrops</code> results is to look at the droplets below our empty threshold. We are assuming that these droplets only contain ambient RNA and therefore the null hypothesis should be true and the distribution of p-values should be approximately uniform.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="01_empty_droplets_files/figure-html/empty-drops-pvals-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
<p>Peaks near zero would tell us that not all of the droplets below the threshold are truly empty and that we should lower it.</p>
<p>We can also plot the negative log-probability against the total counts to see which droplets <code>EmptyDrops</code> has selected.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="01_empty_droplets_files/figure-html/empty-drops-plot-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
<p>There were 1187710 droplets with less than 100 counts which were used to make up the ambient RNA pool. Of the remaining 257418 droplets 45819 were found to have profiles significantly different from the ambient RNA and should contain cells.</p>
</section>
<section id="comparison" class="level1">
<h1>Comparison</h1>
<p>Let’s quickly compare differences between the selection methods using an UpSet plot:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="01_empty_droplets_files/figure-html/compare-overlap-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
<p>We can see that most of the cells are identified by both methods. Our use of the <code>EmptyDrops</code> algorithm has identified even more cells than Cell Ranger v3 but there are very few cells that Cell Ranger identifed which <code>EmptyDrops</code> didn’t.</p>
</section>
<section id="selection" class="level1">
<h1>Selection</h1>
<p>We are going to perform further quality control of these cells anyway so at this stage we will keep those that were selected either by Cell Ranger v3 or by <code>EmptyDrops</code>.</p>
</section>
<section id="summary" class="level1">
<h1>Summary</h1>
<section id="parameters" class="level2">
<h2 class="anchored" data-anchor-id="parameters">Parameters</h2>
<p>This table describes parameters used and set in this document.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="table table-striped caption-top table-sm small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Parameter</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Value</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left; width: 2cm;">n_droplets</td>
<td style="text-align: left; width: 3cm;">1445128</td>
<td style="text-align: left; width: 15cm;">Number o....</td>
</tr>
<tr class="even">
<td style="text-align: left; width: 2cm;">empty_thresh</td>
<td style="text-align: left; width: 3cm;">100</td>
<td style="text-align: left; width: 15cm;">Droplets....</td>
</tr>
<tr class="odd">
<td style="text-align: left; width: 2cm;">emp_iters</td>
<td style="text-align: left; width: 3cm;">60000</td>
<td style="text-align: left; width: 15cm;">Number o....</td>
</tr>
<tr class="even">
<td style="text-align: left; width: 2cm;">emp_fdr</td>
<td style="text-align: left; width: 3cm;">0.01</td>
<td style="text-align: left; width: 15cm;">FDR cuto....</td>
</tr>
<tr class="odd">
<td style="text-align: left; width: 2cm;">n_cellranger</td>
<td style="text-align: left; width: 3cm;">39278</td>
<td style="text-align: left; width: 15cm;">Number o....</td>
</tr>
<tr class="even">
<td style="text-align: left; width: 2cm;">n_empdrops</td>
<td style="text-align: left; width: 3cm;">45819</td>
<td style="text-align: left; width: 15cm;">Number o....</td>
</tr>
<tr class="odd">
<td style="text-align: left; width: 2cm;">n_cells</td>
<td style="text-align: left; width: 3cm;">46155</td>
<td style="text-align: left; width: 15cm;">Number o....</td>
</tr>
</tbody>
</table>


</div>
</div>
</section>
<section id="output-files" class="level2">
<h2 class="anchored" data-anchor-id="output-files">Output files</h2>
<p>This table describes the output files produced by this document. Right click and <em>Save Link As…</em> to download the results (to be implemented).</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">File</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><a href="https///github.com/nami2309/sterile_inflammation//main/output/01_empty_droplets/parameters.json">parameters.json</a></td>
<td style="text-align: left;">Parameters set and used in this analysis</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https///github.com/nami2309/sterile_inflammation//main/output/01_empty_droplets/01_SI_1_barcodes_kept.png">01_SI_1_barcodes_kept.png</a></td>
<td style="text-align: left;">Barcodes kept in CellRanger V3 (PNG)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https///github.com/nami2309/sterile_inflammation//main/output/01_empty_droplets/01_SI_2_empty_droplets_distribution.png">01_SI_2_empty_droplets_distribution.png</a></td>
<td style="text-align: left;">Distribution of empty droplets (PNG)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https///github.com/nami2309/sterile_inflammation//main/output/01_empty_droplets/01_SI_3_barcodes_negative_prob.png">01_SI_3_barcodes_negative_prob.png</a></td>
<td style="text-align: left;">Negative log probability of barcode counts (PNG)</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https///github.com/nami2309/sterile_inflammation//main/output/01_empty_droplets/01_SI_4_compare_methods.png">01_SI_4_compare_methods.png</a></td>
<td style="text-align: left;">Compare both methods (PNG)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https///github.com/nami2309/sterile_inflammation//main/output/01_empty_droplets/01_selected.Robj">01_selected.Robj</a></td>
<td style="text-align: left;">A SingleCellExperiment object of droplets that were called as cells by either Cell ranger v3 or Empty drops</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="session-information" class="level2">
<h2 class="anchored" data-anchor-id="session-information">Session information</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.3 (2024-02-29)
 os       macOS Sonoma 14.6.1
 system   x86_64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Bratislava
 date     2024-12-26
 pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 ! package              * version   date (UTC) lib source
 P abind                  1.4-8     2024-09-12 [?] CRAN (R 4.3.3)
   beachmat               2.18.1    2024-02-14 [1] Bioconductor 3.18 (R 4.3.2)
   Biobase              * 2.62.0    2023-10-24 [1] Bioconductor
   BiocGenerics         * 0.48.1    2023-11-01 [1] Bioconductor
 P BiocManager            1.30.25   2024-08-28 [?] CRAN (R 4.3.3)
   BiocParallel           1.36.0    2023-10-24 [1] Bioconductor
 P bitops                 1.0-9     2024-10-03 [?] CRAN (R 4.3.3)
 P cachem                 1.1.0     2024-05-16 [?] CRAN (R 4.3.3)
   cli                    3.6.3     2024-06-21 [1] RSPM (R 4.3.0)
 P codetools              0.2-19    2023-02-01 [?] CRAN (R 4.3.3)
 P colorspace             2.1-1     2024-07-26 [?] CRAN (R 4.3.3)
 P cowplot              * 1.1.3     2024-01-22 [?] CRAN (R 4.3.2)
 P crayon                 1.5.3     2024-06-20 [?] CRAN (R 4.3.3)
   DelayedArray           0.28.0    2023-10-24 [1] Bioconductor
   DelayedMatrixStats     1.24.0    2023-10-24 [1] Bioconductor
 P devtools               2.4.5     2022-10-11 [?] CRAN (R 4.3.0)
   digest                 0.6.37    2024-08-19 [1] RSPM (R 4.3.0)
 P dplyr                * 1.1.4     2023-11-17 [?] CRAN (R 4.3.0)
 P dqrng                  0.4.1     2024-05-28 [?] RSPM
   DropletUtils         * 1.22.0    2023-10-24 [1] Bioconductor
   edgeR                  4.0.16    2024-02-18 [1] Bioconductor 3.18 (R 4.3.2)
 P ellipsis               0.3.2     2021-04-29 [?] CRAN (R 4.3.0)
   evaluate               1.0.1     2024-10-10 [1] RSPM (R 4.3.0)
 P fastmap                1.2.0     2024-05-15 [?] CRAN (R 4.3.3)
 P forcats              * 1.0.0     2023-01-29 [?] RSPM
   fs                     1.6.5     2024-10-30 [1] RSPM (R 4.3.0)
 P generics               0.1.3     2022-07-05 [?] CRAN (R 4.3.0)
   GenomeInfoDb         * 1.38.8    2024-03-15 [1] Bioconductor 3.18 (R 4.3.3)
   GenomeInfoDbData       1.2.11    2024-12-25 [1] Bioconductor
   GenomicRanges        * 1.54.1    2023-10-29 [1] Bioconductor
 P ggplot2              * 3.5.1     2024-04-23 [?] CRAN (R 4.3.2)
   glue                   1.8.0     2024-09-30 [1] RSPM (R 4.3.0)
 P gridExtra              2.3       2017-09-09 [?] CRAN (R 4.3.0)
 P gtable                 0.3.6     2024-10-25 [?] CRAN (R 4.3.3)
   HDF5Array              1.30.1    2024-03-06 [1] Bioconductor 3.18 (R 4.3.3)
 P here                 * 1.0.1     2020-12-13 [?] CRAN (R 4.3.0)
 P hms                    1.1.3     2023-03-21 [?] CRAN (R 4.3.0)
 P htmltools              0.5.8.1   2024-04-04 [?] CRAN (R 4.3.2)
 P htmlwidgets            1.6.4     2023-12-06 [?] CRAN (R 4.3.0)
 P httpuv                 1.6.15    2024-03-26 [?] CRAN (R 4.3.2)
   IRanges              * 2.36.0    2023-10-24 [1] Bioconductor
   jsonlite               1.8.9     2024-09-20 [1] RSPM (R 4.3.0)
 P kableExtra           * 1.4.0     2024-01-24 [?] RSPM
   knitr                  1.49      2024-11-08 [1] RSPM (R 4.3.0)
 P later                  1.4.1     2024-11-27 [?] RSPM
 P lattice                0.22-5    2023-10-24 [?] CRAN (R 4.3.3)
 P lifecycle              1.0.4     2023-11-07 [?] CRAN (R 4.3.0)
   limma                  3.58.1    2023-10-31 [1] Bioconductor
   locfit                 1.5-9.10  2024-06-24 [1] RSPM (R 4.3.0)
 P lubridate            * 1.9.4     2024-12-08 [?] RSPM
   magrittr               2.0.3     2022-03-30 [1] RSPM (R 4.3.0)
 P Matrix               * 1.6-5     2024-01-11 [?] CRAN (R 4.3.3)
   MatrixGenerics       * 1.14.0    2023-10-24 [1] Bioconductor
 P matrixStats          * 1.4.1     2024-09-08 [?] CRAN (R 4.3.3)
 P memoise                2.0.1     2021-11-26 [?] CRAN (R 4.3.0)
 P mime                   0.12      2021-09-28 [?] CRAN (R 4.3.0)
 P miniUI                 0.1.1.1   2018-05-18 [?] CRAN (R 4.3.0)
 P munsell                0.5.1     2024-04-01 [?] CRAN (R 4.3.2)
 P pillar                 1.10.0    2024-12-17 [?] RSPM
 P pkgbuild               1.4.5     2024-10-28 [?] CRAN (R 4.3.3)
 P pkgconfig              2.0.3     2019-09-22 [?] CRAN (R 4.3.0)
 P pkgload                1.4.0     2024-06-28 [?] RSPM
 P plyr                   1.8.9     2023-10-02 [?] CRAN (R 4.3.0)
 P profvis                0.4.0     2024-09-20 [?] CRAN (R 4.3.3)
 P promises               1.3.2     2024-11-28 [?] RSPM
 P purrr                * 1.0.2     2023-08-10 [?] CRAN (R 4.3.0)
   R.methodsS3            1.8.2     2022-06-13 [1] RSPM (R 4.3.0)
   R.oo                   1.27.0    2024-11-01 [1] RSPM (R 4.3.0)
   R.utils                2.12.3    2023-11-18 [1] RSPM (R 4.3.0)
 P R6                     2.5.1     2021-08-19 [?] CRAN (R 4.3.0)
 P Rcpp                   1.0.13-1  2024-11-02 [?] CRAN (R 4.3.3)
   RCurl                  1.98-1.16 2024-07-11 [1] RSPM (R 4.3.0)
 P readr                * 2.1.5     2024-01-10 [?] RSPM
 P remotes                2.5.0     2024-03-17 [?] RSPM
   renv                   1.0.11    2024-10-12 [1] RSPM (R 4.3.0)
   rhdf5                  2.46.1    2023-11-29 [1] Bioconductor
   rhdf5filters           1.14.1    2023-11-06 [1] Bioconductor
   Rhdf5lib               1.24.2    2024-02-07 [1] Bioconductor 3.18 (R 4.3.2)
   rlang                  1.1.4     2024-06-04 [1] RSPM (R 4.3.0)
   rmarkdown              2.29      2024-11-04 [1] RSPM (R 4.3.0)
 P rprojroot              2.0.4     2023-11-05 [?] CRAN (R 4.3.0)
 P rstudioapi             0.17.1    2024-10-22 [?] CRAN (R 4.3.3)
   S4Arrays               1.2.1     2024-03-06 [1] Bioconductor 3.18 (R 4.3.3)
   S4Vectors            * 0.40.2    2023-11-23 [1] Bioconductor
 P scales                 1.3.0     2023-11-28 [?] CRAN (R 4.3.0)
   scuttle                1.12.0    2023-10-24 [1] Bioconductor
 P sessioninfo            1.2.2     2021-12-06 [?] CRAN (R 4.3.0)
 P shiny                  1.10.0    2024-12-14 [?] RSPM
   SingleCellExperiment * 1.24.0    2023-10-24 [1] Bioconductor
   SparseArray            1.2.4     2024-02-11 [1] Bioconductor 3.18 (R 4.3.2)
   sparseMatrixStats      1.14.0    2023-10-24 [1] Bioconductor
   statmod                1.5.0     2023-01-06 [1] RSPM (R 4.3.0)
 P stringi                1.8.4     2024-05-06 [?] CRAN (R 4.3.2)
   stringr              * 1.5.1     2023-11-14 [1] RSPM (R 4.3.0)
   SummarizedExperiment * 1.32.0    2023-10-24 [1] Bioconductor
 P svglite                2.1.3     2023-12-08 [?] RSPM
 P systemfonts            1.1.0     2024-05-15 [?] RSPM
 P tibble               * 3.2.1     2023-03-20 [?] CRAN (R 4.3.0)
 P tidyr                * 1.3.1     2024-01-24 [?] CRAN (R 4.3.2)
 P tidyselect             1.2.1     2024-03-11 [?] CRAN (R 4.3.2)
 P tidyverse            * 2.0.0     2023-02-22 [?] CRAN (R 4.3.0)
 P timechange             0.3.0     2024-01-18 [?] CRAN (R 4.3.0)
 P tzdb                   0.4.0     2023-05-12 [?] CRAN (R 4.3.0)
 P UpSetR               * 1.4.0     2019-05-22 [?] CRAN (R 4.3.0)
 P urlchecker             1.0.1     2021-11-30 [?] CRAN (R 4.3.0)
 P usethis                3.1.0     2024-11-26 [?] CRAN (R 4.3.3)
 P vctrs                  0.6.5     2023-12-01 [?] CRAN (R 4.3.0)
 P viridisLite            0.4.2     2023-05-02 [?] CRAN (R 4.3.0)
 P withr                  3.0.2     2024-10-28 [?] CRAN (R 4.3.3)
   xfun                   0.49      2024-10-31 [1] RSPM (R 4.3.0)
 P xml2                   1.3.6     2023-12-04 [?] CRAN (R 4.3.0)
 P xtable                 1.8-4     2019-04-21 [?] CRAN (R 4.3.0)
   XVector                0.42.0    2023-10-24 [1] Bioconductor
   yaml                   2.3.10    2024-07-26 [1] RSPM (R 4.3.0)
   zlibbioc               1.48.2    2024-03-13 [1] Bioconductor 3.18 (R 4.3.3)

 [1] /Users/nami/Desktop/sterile_inflammation/renv/library/R-4.3/x86_64-apple-darwin20
 [2] /Users/nami/Library/Caches/org.R-project.R/R/renv/sandbox/R-4.3/x86_64-apple-darwin20/b06620f4

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>